docs/architecture/engine.md

```diff
@@ -0,0 +1,270 @@
+   │     │    │             └───────────┘               │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │               ┌───────┐                 │     │      │
+   │     │    └───────────────┤Shard 1├─────────────────┘     │      │
+   │     │                    └───────┘                       │      │
+   │     │             ┌──────────────────────┐               │      │
+   │     └─────────────┤     Namespace 1      ├───────────────┘      │
+   │                   └──────────────────────┘                      │
+   │                                                                 │
+   │                                                                 │
+   │                                                                 │
+   │              ┌───────────────────────────────┐                  │
+   └──────────────┤           Database            ├──────────────────┘
+                  └───────────────────────────────┘
+```
+
+The in-memory portion of M3DB is implemented via a hierarchy of datastructures:
+
+1. A "database" (one per M3DB node, effectively a singleton)
+
+2. "Namespaces" which are similar to tables or namespaces in other databases. A database "owns" numerous namespaces, and each namespace has a unique name as well as distinct configuration with regards to data retention and blocksize (which we will discuss in more detail later).
+
+3. ["Shards"](sharding.md) which are owned by namespaces. Shards are effectively the same as "virtual shards" in Cassandra in that they provide arbitrary distribution of time series data via a simple hash of the series ID. Shards are useful through the entire M3DB stack in that they make horizontal scaling and adding / removing nodes without downtime trivial at the cluster level, as well as providing more fine grained lock granularity at the memory level, and finally they inform the filesystem organization in that data belonging to the same shard will be used / dropped together and can be kept in the same file.
+
+4. "Series" which are owned by shards. A series is the minimum unit that comes to mind when you think of "time series" data. Ex. The CPU level for a single host in a datacenter over a period of time could be represented as a series with id "<HOST_IDENTIFIER>.system.cpu.utilization" and a vector of tuples in the form of (TIMESTAMP, CPU_LEVEL)
```

Comment:
>A series is the minimum unit that comes to mind when you think of "time series" data."

The sentence is misleading, considering you talk about a smaller unit of time series data right after this paragraph (blocks). Could you rephrase to clarify what you mean about being a single time series unit, as opposed to the data associated with the unit.

Reply:
Also, clarify that the example is only a logical equivalent illustration, and we don't store data like that.

Reply:
will do